src/libsystemd/sd-journal/journal-file.c

```diff
@@ -1659,14 +1659,13 @@ static int journal_file_append_data(
                 const void *data, uint64_t size,
                 Object **ret, uint64_t *ret_offset) {
 
-        uint64_t hash, p;
-        uint64_t osize;
-        Object *o;
+        uint64_t hash, p, fp, osize;
+        Object *o, *fo;
         int r, compression = 0;
         const void *eq;
 
         assert(f);
-        assert(data || size == 0);
+        assert(data && size > 0);
```

Comment:
i figure this should not be an assert, i.e. passing zero length data should be refused politely, not fatally

Reply:
If by the time this function is called, if there are still NULL or zero-sized pieces of data in the iovec, wouldn't that indicate a bigger problem? I figure such validations should be done when we read the data from its source (another journal file, socket, kmsg, ...) and throughout the rest of the code we should assume we don't have zero-sized data (but validate that assumption using asserts)

Reply:
sure, they do indicate a bigger problem, but one we currently don't check sufficiently for afaics. i.e.  journal_file_append_entry() as primary use of journal_file_append_data() doesn't verify that, neither does journal_file_copy_entry(). In particular the latter copies stuff from existing journal files, and they might be rogue, who knows. 

assert() are the right approach for catching *programming* errors, but this appears to be something that is under control of whoever supplies a journal file to us, and in that case we really should generate runtime errors, not assertions hence.

Reply:
I fully agree, that's why I added a size check before we call `journal_file_append_data` in `journal_file_copy_entry` along with this change.

Anyway, it's true that we don't check this in `journal_file_append_entry` so degraded to a runtime error. I kept the iovec assert in `journal_file_append_entry` as that same assert is also done in both call sites of `journal_file_append_entry`.